src/openclon12/resource_tasks.cpp

```diff
@@ -2569,8 +2584,9 @@ class MapCopyTask : public MapTask
 private:
     void RecordImpl() final
     {
-        // Always read back data so we don't write garbage into regions the app didn't write
+        if(m_MapFlags & CL_MAP_READ)

```

Comment:
Skipping this readback step causes the map to effectively change from [`CL_MAP_WRITE`](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_MAP_WRITE) to [`CL_MAP_WRITE_INVALIDATE_REGION`](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_MAP_WRITE_INVALIDATE_REGION). The spec for `CL_MAP_WRITE` says:

> This flag specifies that the region being mapped in the memory object is being mapped for writing.

>
> The pointer returned by [clEnqueueMapBuffer](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clEnqueueMapBuffer) ([clEnqueueMapImage](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clEnqueueMapImage)) is guaranteed to contain the latest bits in the region being mapped when the [clEnqueueMapBuffer](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clEnqueueMapBuffer) ([clEnqueueMapImage](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clEnqueueMapImage)) command has completed

So, unfortunately `WRITE` must be more or less identical to `READ | WRITE`. I'm open to changing the heap flags; apps shouldn't be doing CPU writes, and so using `HEAP_TYPE_UPLOAD` (`WRITE_COMBINE`) instead of `READBACK` (`WRITE_BACK`) for the CPU side of things is a reasonable change/optimization.

Reply:
ack. Pls review my latest force push.
Also d3d12translationlayer\ImmediateContext.cpp line 1744 assert might need to change.
In my commit, A MAP_WRITE mappable resource will be convert to COPY_DEST while being mapped. which will be rejected in the assert if we are running in debug build.
If I want to make change to this code, should the code change be made in this repo or in D3D12TranslationLayer and later being sync to this repo?

Reply:
> If I want to make change to this code, should the code change be made in this repo or in D3D12TranslationLayer and later being sync to this repo?

In https://github.com/microsoft/OpenCLOn12/pull/57, I pulled in a copy of the D3D12TranslationLayer dependency and made significant changes to that copy, so we are no longer depending on / tracking the primary repository there. I just didn't fully remove the namespacing, but there's no reason that couldn't be done.